### PR TITLE
Setup test framework

### PR DESCRIPTION
**Pull request by alex-ju**
This PR sets up the test framework using [The GDS Way](https://github.com/maxcbc/gds-way/blob/600f30643be4438baf1b52d13d9f0896f2bfaf29/source/manuals/programming-languages/nodejs/testing.html.md.erb) recommended tools as follows:
- prepares `app.js` to be tested
- updates documentation on testing
- splits testing into 3 types:
  - App testing (Node.js/Express) `test-app`
  - Components testing `test-components`
  - Automation testing (Gulp tasks) `test-tasks`
- Adds helper functions to `/lib` for reading and comparing directories/files
- Adds testing tools dependencies
  - [Mocha](https://github.com/mochajs/mocha) as a test runner
  - [Chai](https://github.com/chaijs/chai) as assertion library
  - [Cheerio](https://github.com/cheeriojs/cheerio) for DOM parsing and manipulation
  - [Request](https://github.com/request/request) for HTTP requests
- Creates gulp tasks for testing the Node.js app (`test:app`) and the Gulp tasks (`test:tasks`)
- Adds basic app test to ensure `/` returns a list with all components
- Adds tasks tests ([Trello ticket](https://trello.com/c/hHnVtwtk)) to ensure:
  - All components are being copied to `dist` and `packages`
  - All the right files are being copied to `dist` and `packages` for each component

***Note to reviewer:** Some tests are failing. PRs will follow to fix the errors.*

### CODE REVIEW

File: docs/development-and-publish-tasks.md [hunk 1]
```diff
@@ -59,6 +59,18 @@ This task will:
 This task will:
 - will run accessibility test (tenon and axe)
 
+**`gulp test:app`**
+
+This task will:
```
Comment (36degrees) [inline docs/development-and-publish-tasks.md]: Duplication of 'will' – which I notice is also a problem above! Would be good to fix this throughout.

File: lib/helper-functions.js [hunk 2]
```diff
@@ -18,5 +21,36 @@ const componentNameToMacroName = componentName => {
...
+
+// Read components from various sources
+exports.SrcComponentList = fs.readdirSync(configPaths.src + 'components/')
+exports.DistComponentList = fs.readdirSync(configPaths.dist + 'components/')
+exports.PackagesComponentList = fs.readdirSync(configPaths.packages)
+
+// Compare component files from src with dist
+const compareComponentDist = (componentName) => {
+  let distComponentFiles = fs.readdirSync(configPaths.dist + 'components/' + componentName)
+  // We don't want to include the yaml file in dist
```
Comment (36degrees) [inline lib/helper-functions.js]: or the readme template?

File: lib/helper-functions.js [hunk 3]
```diff
@@ -18,5 +21,36 @@ const componentNameToMacroName = componentName => {
...
-
 exports.componentNameToMacroName = componentNameToMacroName
+
+// Read components from various sources
+exports.SrcComponentList = fs.readdirSync(configPaths.src + 'components/')
+exports.DistComponentList = fs.readdirSync(configPaths.dist + 'components/')
+exports.PackagesComponentList = fs.readdirSync(configPaths.packages)
+
+// Compare component files from src with dist
+const compareComponentDist = (componentName) => {
```
Comment (36degrees) [inline lib/helper-functions.js]: I wonder about the naming of this function – it's not actually doing any 'comparison', right? It's returning an object with two arrays, that you can then compare back in the test.

Should it even be one function? I think we could split it into e.g.

```
const distFilesForComponent(componentName) => {
   return fs.readdirSync(configPaths.dist + 'components/' + componentName)
}

const expectedDistFilesForComponent(componentName) => {
  let srcComponentFiles = fs.readdirSync(configPaths.src + 'components/' + componentName)
                        .filter(file => (file !== componentName + '.yaml' && file !== 'index.njk'))
  // We expect dist to contain an html sample
  srcComponentFiles.push(componentName + '.html')
  return srcComponentFiles.sort()
}
```

Then the test would be something like:
```
let actualDistFiles = lib.distFilesForComponent(componentName)
let expectedDistFiles = lib.expectedDistFilesForComponent(componentName)
expect(actualDistFiles).to.have.same.members(expectedDistFiles)
```

Which I think is clearer?

File: test/app.spec.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/* eslint-env mocha */
+const expect = require('chai').expect
+const request = require('request')
+const cheerio = require('cheerio')
+
+const app = require('../app.js')
+const lib = require('../lib/helper-functions')
+
+const serverOptions = {
```
Comment (36degrees) [inline test/app.spec.js]: I'm unclear what this is doing – is it options used to create a server used for the test, or parameters used when making a request?

File: test/tasks.spec.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+      // Check component files in packages
+      it('should copy the right files to packages', () => {
+        let c = lib.compareComponentPackage(componentName)
+        expect(c.packagesComponentFiles).to.have.same.members(c.srcComponentFiles)
+      })
+    })
+  }
+
+  for (let i = 0; i < lib.SrcComponentList.length; i++) {
```
Comment (36degrees) [inline test/tasks.spec.js]: Could this be done with a `forEach`?

File: lib/helper-functions.js (hunk 3 above)
Comment (alex-ju) [reply lib/helper-functions.js]: Sounds good. Will update.

File: test/app.spec.js (hunk 4 above)
Comment (alex-ju) [reply test/app.spec.js]: parameters used when making a request. I should probably make that more obvious through naming

File: test/tasks.spec.js (hunk 5 above)
Comment (alex-ju) [reply test/tasks.spec.js]: `for` is faster, `forEach` easier to read. If we're fine with `forEach` and use it consistently I can comply.

File: lib/helper-functions.js (hunk 2 above)
Comment (alex-ju) [reply lib/helper-functions.js]: Yes, if we don't want `Readme.md`s in `dist` I can exclude them. I was looking at what the tasks are supposed to be doing.

Comment (36degrees) [reply lib/helper-functions.js]: Sorry, the readme _template_ (`index.njk`) – which is not mentioned in this comment.

File: test/tasks.spec.js (hunk 5 above)
Comment (36degrees) [reply test/tasks.spec.js]: I don't have strong opinions, I just worry that this bit of the test is hard to read at the minute. I'd also be tempted to rename `_Published` to make it clearer what's going on – e.g. `defineTestsForComponent` or something.

**Comment by alex-ju** [comment]
@36degrees updated, please have a look when you get time.

File: lib/file-helper.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+}
+exports.expectedDistFilesForComponent = expectedDistFilesForComponent
+
+const packagesFilesForComponent = componentName => {
+  return fs.readdirSync(configPaths.packages + componentName)
+}
+exports.packagesFilesForComponent = packagesFilesForComponent
+
+const expectedPackagesFilesForComponent = componentName => {
+  // We don't want to include the yaml file and index.njk in dist
```
Comment (36degrees) [inline lib/file-helper.js]: in **packages**?

File: lib/file-helper.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+const fs = require('fs')
+const configPaths = require('../config/paths.json')
+
+// Read components from various sources
+exports.SrcComponentList = fs.readdirSync(configPaths.src + 'components/')
+exports.DistComponentList = fs.readdirSync(configPaths.dist + 'components/')
+exports.PackagesComponentList = fs.readdirSync(configPaths.packages)
+
+const distFilesForComponent = componentName => {
```
Comment (36degrees) [inline lib/file-helper.js]: I wonder if these functions would benefit from doc blocks to explain what they're doing and how they're used?

**Comment by 36degrees** [comment]
Would like make sure @hannalaakso has the opportunity to review this as well.

File: app.js [hunk 8]
```diff
@@ -171,3 +171,5 @@ app.use(function (err, req, res, next) {
   res.status(err.status || 500)
   res.render('http-error', { error: 'Internal server error', message: err })
 })
+
+module.exports = server
```
Comment (hannalaakso) [inline app.js]: Is `server` used in `app.spec.js`? It looks like it might be defined again in that file.

File: lib/file-helper.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+// This helper function takes a component name and returns the expected files
+// to be found in the packages folder of that component
+const expectedPackagesFilesForComponent = componentName => {
+  // We don't want to include the yaml file and index.njk in dist
+  let srcComponentFiles = fs.readdirSync(configPaths.src + 'components/' + componentName)
+                        .filter(file => (file !== componentName + '.yaml' && file !== 'index.njk'))
+  // We expect the package to contain a package.json and an html sample
+  srcComponentFiles.push('package.json')
+  srcComponentFiles.push(componentName + '.html')
+  srcComponentFiles.sort()
```
Comment (hannalaakso) [inline lib/file-helper.js]: This seems to sort the files twice.

File: lib/file-helper.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+//
+// This helper function takes a component name and returns the expected files
+// to be found in the dist/components folder of that component
+const expectedDistFilesForComponent = componentName => {
+  // We don't want to include the yaml file and index.njk in dist
+  let srcComponentFiles = fs.readdirSync(configPaths.src + 'components/' + componentName)
+                        .filter(file => (file !== componentName + '.yaml' && file !== 'index.njk'))
+  // We expect dist to contain an html sample
+  srcComponentFiles.push(componentName + '.html')
+  return srcComponentFiles.sort()
```
Comment (hannalaakso) [inline lib/file-helper.js]: We're sorting the files here and in `tasks.spec.js`. I wonder if sorting is needed as `to.have.same.members` in Chai doesn't seem to check order, there's a separate `to.have.ordered.members`.

File: lib/file-helper.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+// to be found in the packages folder of that component
+const expectedPackagesFilesForComponent = componentName => {
+  // We don't want to include the yaml file and index.njk in dist
+  let srcComponentFiles = fs.readdirSync(configPaths.src + 'components/' + componentName)
+                        .filter(file => (file !== componentName + '.yaml' && file !== 'index.njk'))
+  // We expect the package to contain a package.json and an html sample
+  srcComponentFiles.push('package.json')
+  srcComponentFiles.push(componentName + '.html')
+  srcComponentFiles.sort()
+  return srcComponentFiles.sort()
```
Comment (hannalaakso) [inline lib/file-helper.js]: This seems to sort the files twice.

File: app.js (hunk 8 above)
Comment (alex-ju) [reply app.js]: We're exporting the server and import it in the spec as `app` on the following line
`const app = require('../app.js')`

we then save the reference in a local variable and start the server:
`server = app.listen(3000)`

and we're doing that to be able to stop the server after finishing the tests:
`server.close()`

File: lib/file-helper.js (hunk 10 above)
Comment (alex-ju) [reply lib/file-helper.js]: There is no need to sort the files when looking for the same members, but it makes it a lot easier to figure the differences between ` + expected` and `- actual` when tests are failing.

File: lib/file-helper.js (hunk 11 above)
Comment (alex-ju) [reply lib/file-helper.js]: Yep, sorted twice. Fixed that now.

File: app.js (hunk 8 above)
Comment (hannalaakso) [reply app.js]: Cool that makes sense. Thanks for explaining 👍

File: test/tasks.spec.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    // Check to have all the components in dist and nothing else
+    it('should copy all components to dist', () => {
+      expect(lib.DistComponentList).to.deep.equal(lib.SrcComponentList)
+    })
+
+    // Check to have all the components (plus all, globals, icons) in packages and nothing else
+    it('should copy all components plus "all", "globals" and "icons" to packages', () => {
+      let expectedPackagesComponentList = lib.SrcComponentList.slice()
+      expectedPackagesComponentList.push('all', 'globals', 'icons')
+      expectedPackagesComponentList.sort()
```
Comment (hannalaakso) [inline test/tasks.spec.js]: Could we add a comment here to say that there's no need to sort the files for `to.have.same.members` but it makes debugging failing tests a lot easier when the files are sorted?

File: lib/file-helper.js (hunk 10 above)
Comment (hannalaakso) [reply lib/file-helper.js]: Aha that makes sense. I've left a comment below in the code to explain this.

File: test/tasks.spec.js (hunk 12 above)
Comment (alex-ju) [reply test/tasks.spec.js]: Sure. Comments added.

**Pull request merged**